providers/apis/defi/raydium/provider.go

```diff
@@ -0,0 +1,252 @@
+	market oracletypes.ProviderMarketMap
+
+	// config is the APIConfiguration for this provider
+	config config.APIConfig
+
+	// client is the solana JSON-RPC client used to query the API.
+	client SolanaJSONRPCClient
+
+	// metaDataPerTicker is a map of ticker.String() -> TickerMetadata
+	metaDataPerTicker map[string]TickerMetadata
+
+	// logger
+	logger *zap.Logger
+}
+
+// NewAPIPriceFetcher returns a new APIPriceFetcher. This method constructs the
+// default solana JSON-RPC client in accordance with the config's URL param.
+func NewAPIPriceFetcher(
+	market oracletypes.ProviderMarketMap,
+	config config.APIConfig,
+	logger *zap.Logger,
+) (*APIPriceFetcher, error) {
+	return NewAPIPriceFetcherWithClient(
+		market,
+		config,
+		rpc.New(config.URL),
+		logger,
+	)
+}
+
+// NewAPIPriceFetcherWithClient returns a new APIPriceFetcher. This method requires
+// that the given market + config are valid, otherwise a nil implementation + an error
+// will be returned.
+func NewAPIPriceFetcherWithClient(
+	market oracletypes.ProviderMarketMap,
+	config config.APIConfig,
+	client SolanaJSONRPCClient,
+	logger *zap.Logger,
+) (*APIPriceFetcher, error) {
+	if err := config.ValidateBasic(); err != nil {
+		return nil, err
+	}
+
+	if err := market.ValidateBasic(); err != nil {
+		return nil, err
+	}
+
+	// check fields of config
+	if config.Name != Name {
+		return nil, fmt.Errorf("config.Name is not %s", Name)
+	}
+
+	if market.Name != Name {
+		return nil, fmt.Errorf("market.Name is not %s", Name)
+	}
+
+	if !config.Enabled {
+		return nil, fmt.Errorf("config is not enabled")
+	}
+
+	// generate metadata per ticker
+	metadataPerTicker := make(map[string]TickerMetadata)
+	for _, ticker := range market.OffChainMap {
+		metadata, err := unmarshalMetadataJSON(ticker.Metadata_JSON)
+		if err != nil {
+			return nil, fmt.Errorf("error unmarshalling metadata for ticker %s: %w", ticker.String(), err)
+		}
+
+		if err := metadata.ValidateBasic(); err != nil {
+			return nil, fmt.Errorf("metadata for ticker %s is invalid: %w", ticker.String(), err)
+		}
+
+		metadataPerTicker[ticker.String()] = metadata
+	}
+
+	return &APIPriceFetcher{
+		market:            market,
+		config:            config,
+		client:            client,
+		metaDataPerTicker: metadataPerTicker,
+		logger:            logger.With(zap.String("provider", Name)),
+	}, nil
+}
+
+// FetchPrices fetches prices from the solana JSON-RPC API for the given currency-pairs. Specifically
+// for each currency-pair,
+//   - Query the raydium API base (coin) / quote (pc) token vault addresses
+//   - Normalize the token balances by 1e18
+//   - Calculate the price as quote / base, and scale by ticker.Decimals
+func (pf *APIPriceFetcher) FetchPrices(
+	ctx context.Context,
+	tickers []mmtypes.Ticker,
+) providertypes.GetResponse[mmtypes.Ticker, *big.Int] {
+	// get the acounts to query in order of the tickers given
+	accounts := make([]solana.PublicKey, len(tickers)*2)
```

Comment:
`len(tickers)*2` is used multiple times.

let's assign to a variable with a nice name